fix(forms): Correct required validator to handle objects with a property `length: 0`

s-gbz opened the pull request:
Form required validator should not reject objects that contain a length attribute set to zero.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Issue Number: #30718

{id: 1, length: 0, width: 0} is evaluated as empty value.

## What is the new behavior?
{id: 1, length: 0, width: 0} is evaluated as **non-empty** value.

## Does this PR introduce a breaking change?

- [X] Yes (Didn't back when it was submitted)
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

s-gbz commented:
Hello @kara can you take a look at the changes?
I'm not quite sure why ci fails to build the npm package, but I don't think it's related to the changes made.

PS: Sorry for the messy commits! I'm working on a submachine and had a bit of trouble :)

s-gbz commented:
Hey there @kapunahelewong!
Is there any chance you could refer this issue to someone to take a look?
I've been dragging it to long and think we're almost done here :)

Please excuse in case I am too unpatient!

kapunahelewong commented on packages/forms/src/validators.ts:
```diff
@@ -13,8 +13,13 @@ import {AsyncValidatorFn, ValidationErrors, Validator, ValidatorFn} from './dire
+  /**
+   * We check if the object is a string or array before evaluating the length attribute.
+   * We thereby avoid falsly rejecting objects that contain a custom length attribute.
+   * E.g. object {id: 1, length: 0, width: 0} should not be returned as empty.
```
E.g. --> For example

We're trying to switch any instances of Latin abbreviations to their English counterparts for clarity.

kapunahelewong commented on packages/forms/src/validators.ts:
```diff
@@ -13,8 +13,13 @@ import {AsyncValidatorFn, ValidationErrors, Validator, ValidatorFn} from './dire
-  // we don't check for string here so it also works with arrays
-  return value == null || value.length === 0;
+  /**
+   * We check if the object is a string or array before evaluating the length attribute.
```
Remove "we" and start the sentence with "Check". We're trying to reduce the use of "we".

kapunahelewong commented on packages/forms/src/validators.ts:
```diff
@@ -13,8 +13,13 @@ import {AsyncValidatorFn, ValidationErrors, Validator, ValidatorFn} from './dire
-  return value == null || value.length === 0;
+  /**
+   * We check if the object is a string or array before evaluating the length attribute.
+   * We thereby avoid falsly rejecting objects that contain a custom length attribute.
```
Same note about "we" here. This one could instead be, "This avoids falsely...."

(make sure to catch the typo falsly --> falsely)

kapunahelewong reviewed (commented):
Hi @s-gbz! Thanks for reaching out. I've reviewed the comments for style, but we need someone to review your code changes. I'll see if I can get someone to help in that regard. In the meantime, if more than a day or so passes, will you tag me again so I can help you keep this moving?

s-gbz commented:
Hey there @kapunahelewong :) Thanks a lot for the review, I'll get the changes worked in by tomorrow.
Have you found someone appropriate? (And who might as well review #33572 ?)

kapunahelewong commented:
Hi @s-gbz! Let me reach out again. For https://github.com/angular/angular/pull/33572, you might like to change it from draft so that the codeowners are requested. Otherwise, though I'll check in with my team lead to make sure they go the right direction.

kapunahelewong commented:
Hey, @s-gbz, I've added the `PR action: review` label so it is in the queue and the right people will definitely see it.

s-gbz commented:
The original commits were correct. I hope (guess?) we're good to go if the issue isn't outdated yet @aikidave .

```
Examining 3 commit(s) in the provided range: 822b838fbcb3dfaaec8b8e800dbbdd903fbfc100..2a51a21757dfeb7b6bbf856d80148a7ece4742b9
✘  Invalid commit message
fix(forms): fixup! comment change request
  Error:
    The commit message body does not meet the minimum length of 20 characters
  
  The expected format for a commit is: 
  <type>(<scope>): <summary>
  
  <body>
```

AndrewKushnir commented:
Hi @s-gbz, the `lint` failure was caused by the fact that some commits don't have a commit body. It looks like you can just merge all of them into one and make sure there is a commit message body included. Also I wanted to mention that this PR contains a breaking change. I've added it to the `v12-candidates` milestone so that it can be included into prioritization once we start working on the v12 release. Thank you.

kapunahelewong commented on packages/forms/src/validators.ts:
```diff
@@ -14,8 +14,13 @@ import {AsyncValidator, AsyncValidatorFn, ValidationErrors, Validator, Validator
+  /**
+   * Check if the object is a string or array before evaluating the length attribute.
+   * This avoids falsely rejecting objects that contain a custom length attribute.
+   * For example object {id: 1, length: 0, width: 0} should not be returned as empty.
```
```suggestion
   * For example, the object {id: 1, length: 0, width: 0} should not be returned as empty.
```

kapunahelewong reviewed (approved):
Just a small comment. Thank you, @s-gbz! Otherwise, LGTM!

Reviewed-for: global-docs-approvers

s-gbz committed:
fix(forms): Correct empty validator to handle objects with a property `length: 0`.

Form required validator should not reject objects that contain a length attribute set to zero.

Fixes #30718.

Co-authored-by: Dylan Hunn <dylhunn@gmail.com>

BREAKING CHANGE: objects with a length key set to zero will no longer validate as empty.

This is technically a breaking change, since objects with a key `length` and value `0` will no longer validate as empty. This is a very minor change, and any reliance on this behavior is probably a bug anyway.

s-gbz commented:
@kapunahelewong looks good, thanks a lot for your suggestions :) 
Sorry for the many pushes. I forgot to rebase on the upstream branch.
Hope we're good to go now!

kapunahelewong commented:
It's all great! Though the CLA has switched to no. Could you have possibly worked with a different email from the one that you submitted the PR with originally?

s-gbz commented:
I think you have to approve as well. See the google bot :)

s-gbz commented:
@googlebot I consent.

kapunahelewong commented:
Thanks, for re-doing the consent. I tried amending the commit author since that usually works to remove me but this time it's sticking. Glad it's sorted out. :)

alxhub commented on packages/forms/src/validators.ts:
```diff
@@ -14,8 +14,13 @@ import {AsyncValidator, AsyncValidatorFn, ValidationErrors, Validator, Validator
+   * This avoids falsely rejecting objects that contain a custom length attribute.
+   * For example, the object {id: 1, length: 0, width: 0} should not be returned as empty.
+   */
+  return value == null || (typeof value === 'string' && value.length === 0) ||
```
Why not:

```javascript
value == null || ((typeof value === 'string' || Array.isArray(value)) && value.length === 0)
```

s-gbz replied:
Hey there @alxhub unfortunately I can't remember since it's been almost 3 years :smile: 
We could however commit your suggestion if you'd like.
It was probably just a typo that slipped in.

dylhunn replied:
These should be equivalent, Alex's requires a bit less code though.

dylhunn replied:
Pushed a fixup.

dylhunn commented:
I went ahead and rebased this, updated the commit message, and made Alex's fix.
I have also started a TGP, since this is technically a breaking change (although a tiny one): [link](https://test.corp.google.com/OCL:426292603:BASE:426292668:1643944755713:59d31657
)

dylhunn reviewed (approved):
reviewed-for: fw-forms

dylhunn commented:
The failure in `components-repo-unit-tests` is a timeout for the target `//src/material/select:unit_tests_firefox`. Almost certainly a flake, especially given the clean TGP.

dylhunn commented:
This PR was merged into the repository by commit 8dd3f82f946bae86dc6c678d8694ab73d915bbfa.
